subiquity/server/tests/test_server.py

```diff
@@ -232,6 +245,68 @@ async def test_autoinstall_validation__no_error_report(self):
         error = NonReportableError.from_exception(exception)
         self.assertEqual(error, self.server.nonreportable_error)
 
+    @patch("subiquity.server.server.log")
+    async def test_autoinstall_validation__strict_top_level_keys(self, log_mock):
+        """Test strict top-level key enforcement"""
+
+        # Reset base schema
+        self.server.base_schema = SubiquityServer.base_schema
+
+        bad_ai_data = {
+            "version": 1,
+            "interactive-sections": ["identity"],
+            "apt": "Invalid but deferred",
+            "literally-anything": "lmao",
+        }
+        stripped_ai_data = copy.copy(bad_ai_data)
+        stripped_ai_data.pop("literally-anything")
+
+        self.server.autoinstall_config = bad_ai_data
+
+        self.pseudo_load_controllers()
+
+        # OK in Version 1 but ensure warnings and stripped config
+        self.server.validate_autoinstall()
+        log_mock.warning.assert_called()
+        log_mock.error.assert_not_called()
+        self.assertEqual(self.server.autoinstall_config, stripped_ai_data)
+
+        # Not OK in Version >= 2
+        log_mock.reset_mock()
```

Comment:
I guess I kind of don't trust `reset_mock` - it's tough to convince myself that everything that matters has been reset - are the later tests only passing because of retained state from earlier?  What do you think about splitting this test in two parts to cover your version=1 and version=2 cases?  Common setup logic could be refactored.